### PR TITLE
Oob check

### PR DESCRIPTION
**Pull request by ltqin**
oob check using attribute of transform

### CODE REVIEW

**Review by whchung — APPROVED** [review]
LGTM. I can see `hasPadding` has been replaced.

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 1]
```diff
@@ -598,6 +606,10 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
           targetGemmDim2Attr.push_back(b.getNamedAttr(
               "names", b.getArrayAttr({b.getStringAttr(gemmKPad_name)})));
         }
+        // filter of forward, gemmK=c*y*x
+        filterOobCheckDims.insert(nameToDims["c"]);
+        filterOobCheckDims.insert(nameToDims["y"]);
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:611]: @ltqin can we use `needExtraPad` and `gemmKExtra` to help us understand should `c/y/x` all be inserted, or just `c`?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 2]
```diff
@@ -1308,6 +1352,10 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
           targetGemmDim2Attr.push_back(b.getNamedAttr(
               "names", b.getArrayAttr({b.getStringAttr(gemmKPad_name)})));
         }
+        // input of forward, gemmK = ci * y * x( y x from hi wi)
+        inputOobCheckDims.insert(nameToDims["ci"]);
+        inputOobCheckDims.insert(nameToDims["hi"]);
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:1357]: @ltqin can we use `needExtraPad` and `gemmKExtra` to help us understand should `ci/hi/wi` all be inserted, or just `ci`?

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h [hunk 3]
```diff
@@ -1640,6 +1703,10 @@ struct Conv2DRewritePattern : public OpRewritePattern<T> {
           targetGemmDim2Attr.push_back(b.getNamedAttr(
               "names", b.getArrayAttr({b.getStringAttr(gemmKPad_name)})));
         }
+        // output of forward, gemmK = no * ho * wo
+        outputOobCheckDims.insert(nameToDims["no"]);
+        outputOobCheckDims.insert(nameToDims["ho"]);
```
Comment (whchung) [inline mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:1708]: @ltqin can we use `needExtraPad` and `gemmKExtra` to help us understand should `no/ho/wo` all be inserted, or just `no`?

**Review by whchung — DISMISSED** [review]
`miopen.threadwise_copy_v2` also need to be changed.

**Review by whchung — APPROVED** [review]
@ltqin please check my questions and submit these possible optimizations in the next PR.

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 1 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:611]: For gkcxy data layout, if  y =( UpIndex % (Y*X))/X and x = ( UpIndex % (Y*X))%X , I think it not necessary to oob check y and x.
Yes ,"needExtraPad" and "gemmKExtra" can help us understand that  in different case in source code

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 2 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:1357]: same above

File: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h (hunk 3 above)
Comment (ltqin) [reply mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:1708]: same above

**Pull request merged**